extensions/amp-cxense-player/0.1/amp-cxense-player.js

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-cxense-player-0.1.css';
+import {getLengthNumeral, isLayoutSizeDefined} from '../../../src/layout';
+import {loadPromise} from '../../../src/event-helper';
+import {addParamsToUrl} from '../../../src/url';
+import {createLoaderElement} from '../../../src/loader';
+import {getDataParamsFromAttributes} from '../../../src/dom';
+import {setStyles} from '../../../src/style';
+
+const cxDefaults = {
+  apiHost: 'https://api.widget.cx',
+  distEmbedHost: 'https://embed.widget.cx',
+  debugEmbedHost: 'https://stage-embed.widget.cx',
+  distEmbedApp: '/app/player/m4/dist/',
+  debugEmbedApp: '/app/player/m4/debug/',
+    // can't access the window.top.location.href from the iframe to enable sharing the top url
+  attrs: {
+    'share.enable': false,
+  },
+};
+
+class AmpCxense extends AMP.BaseElement {
+
+    /** @override */
+    preconnectCallback(onLayout) {
+      this.preconnect.url(cxDefaults.apiHost, onLayout);
+    }
+
+    /** @override */
+    isLayoutSupported(layout) {
+      return isLayoutSizeDefined(layout);
+    }
+
+    /** @override */
+    renderOutsideViewport() {
+      return false;
+    }
+
+    /** @override */
+    buildCallback() {
+      this.element.classList.add('amp-cxense-player');
+
+        /** @private @const {object} */
+      this.cxDefaults_ = JSON.parse(JSON.stringify(cxDefaults));
+        /** @private @const {object} */
+      this.attrs_ = this.getDataAttributes_();
+
+      this.cxDefaults_.embedHost = this.attrs_.debug
+            ? this.cxDefaults_.debugEmbedHost : this.cxDefaults_.distEmbedHost;
+      this.cxDefaults_.embedApp = this.attrs_.debug
+            ? this.cxDefaults_.debugEmbedApp : this.cxDefaults_.distEmbedApp;
+
+        /** @private @const {number} */
+      this.width_ = getLengthNumeral(this.element.getAttribute('width'));
+        /** @private @const {number} */
+      this.height_ = getLengthNumeral(this.element.getAttribute('height'));
+
+      if (!this.getPlaceholder()) {
+        this.buildImagePoster_();
+        this.loading_(true);
```

Comment:
well, at first, the loader wasn't working for me until I added this tag name to this hardcoded list of [`tagNames`](https://github.com/ampproject/amphtml/blob/7edc5be26488b526db220a7b54f7c1b24d4856f0/src/layout.js#L80-L91)
```javascript
/**
 * Elements that the progess can be shown for. This set has to be externalized
 * since the element's implementation may not be downloaded yet.
 * @enum {boolean}
 * @private  Visible for testing only!
 */
export const LOADING_ELEMENTS_ = {
  'AMP-ANIM': true,
  'AMP-BRIGHTCOVE': true,
  'AMP-EMBED': true,
  'AMP-IFRAME': true,
  'AMP-IMG': true,
  'AMP-INSTAGRAM': true,
  'AMP-LIST': true,
  'AMP-PINTEREST': true,
  'AMP-VIDEO': true,
  'AMP-YOUTUBE': true,
};
```
which is essentially used by [`isLoadingAllowed `](https://github.com/ampproject/amphtml/blob/7edc5be26488b526db220a7b54f7c1b24d4856f0/src/layout.js#L267) and from the comment, it doesn't look like the way to go is to just keep adding tagNames in here